scene/resources/particles_material.cpp

```diff
@@ -261,6 +269,195 @@ void ParticlesMaterial::_update_shader() {
 	code += "}\n";
 	code += "\n";
 
+	//functions to create simplex noise, curl noise in 3D and 4D
+	code += "\n\n";
+	code += "vec4 permute(vec4 x) {\n";
+	code += "	return mod(((x * 34.0) + 1.0) * x, 289.0);\n";
+	code += "}\n";
+	code += "\n";
+
+	code += "\n\n";
+	code += "vec4 taylor_inv_sqrt(vec4 r) {\n";
+	code += "return 1.79284291400159 - 0.85373472095314 * r;\n";
+	code += "}\n";
+	code += "\n";
+
+	code += "\n\n";
+	code += "float permute_float(float x) {\n";
+	code += "	return floor(mod(((x * 34.0) + 1.0) * x, 289.0));\n";
+	code += "}\n";
+
+	code += "\n\n";
+	code += "float taylor_inv_sqrt_float(float r) {\n";
+	code += "	return 1.79284291400159 - 0.85373472095314 * r;\n";
+	code += "}\n";
+
+	code += "\n\n";
+	code += "float snoise_3d(vec3 v) {\n";
+	code += "	vec2 C = vec2(1.0 / 6.0, 1.0 / 3.0) ;\n";
+	code += "	vec4 D = vec4(0.0, 0.5, 1.0, 2.0);\n";
+	code += "	vec3 i = floor(v + dot(v, vec3(C.y)) );\n";
+	code += "	vec3 x0 = v - i + dot(i, vec3(C.x)) ;\n";
+	code += "	vec3 g = step(x0.yzx, x0.xyz);\n";
+	code += "	vec3 l = 1.0 - g;\n";
+	code += "	vec3 i1 = min(g.xyz, l.zxy);\n";
+	code += "	vec3 i2 = max(g.xyz, l.zxy);\n";
+	code += "	vec3 x1 = x0 - i1 + vec3(C.x);\n";
+	code += "	vec3 x2 = x0 - i2 + vec3(C.y);\n";
+	code += "	vec3 x3 = x0 - vec3(D.y); \n";
+	code += "	i = mod(i, 289.0); \n";
+	code += "	vec4 p = permute( permute( permute( \n";
```

Comment:
Throughout the code, I also notice you have some rather inconsistent spacing. Sometimes you'll put a space after the last parameter but before the `)`, and in a couple instances here I notice you put a space between the last `)` and the `;`.

While I don't think it's necessary to fix it, any particular reason for it? Is it for readability?

Reply:
It is my mistake. There is no particular reason for it. I will fix it. Thank you so much